src/Yesod/Page/RenderedRoute.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Yesod.Page.RenderedRoute
+  ( RenderedRoute
+  , getRenderedRoute
+  , updateQueryParameter
+  )
+where
+
+import Data.Aeson
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (fromMaybe)
+import Data.Text (Text, intercalate)
+import UnliftIO (throwString)
+import Yesod.Core
+  ( HandlerSite
+  , MonadHandler
+  , RenderRoute
+  , getCurrentRoute
+  , getRequest
+  , renderRoute
+  , reqGetParams
+  )
+
+-- | Information about the initial route
+data RenderedRoute = RenderedRoute
+  { renderedRouteDomain :: Maybe Text
+  , renderedRoutePath :: [Text]
+  , renderedRouteQuery :: [(Text, Text)]
+  }
+
+instance ToJSON RenderedRoute where
+  -- | Renders as a @'Text'@ URL
+  --
+  -- FIXME: This is wildly naive and we should move to a proper library to
+  -- ensure escaping is handled, but it's meant only to prove the approach and
+  -- pass the current test suite.
+  --
+  toJSON RenderedRoute {..} = String
+    $ fromMaybe "" renderedRouteDomain
+    <> intercalate "/" renderedRoutePath
+    <> renderQuery renderedRouteQuery
+   where
+    renderQuery [] = ""
+    renderQuery qs = "?" <> intercalate "&" (map (\(k, v) -> k <> "=" <> v) qs)
+
+-- | Get the current route as a @'RenderedRoute'@
+--
+-- TODO: supplying the domain is a transitional thing to ease the diff with
+-- previous code, we can just grab Host off the @'Request'@ here, I think.
+--
+-- That would:
+--
+-- 1. Simplify @'PageConfig'@ further
```

Comment:
I actually think I went with `PageConfig` instead of grabbing `Host` because I've seen that bite people. For example ClassLink allows you to access a specific One Roster server through a proxy URL. The `Link` headers that are returned from that request end up encoding the internal domain of the host server, instead of the "pretty" domain of the proxy. However that may be an edge case that isn't worth fighting.

Reply:
In the proxy case I think there's usually another header we could look for (and fall back to `Host`) to get the pretty-or-real value and cover this case. Some proxies don't add the header, but I think that would be even more of an edge-case not to fight with.

I'm also fine keeping this aspect as-is if you prefer it.

Reply:
Meh, I'm fine with using `Host`, but let's follow your plan and kick that to another PR. Just providing some context.